File: 10533 Digit Primes/solution.c
#include <stdio.h>
#define MAX_NUM 999999
#define GET(n) (mark[(n) >> 5] & (1u << ((n) & 31)))
#define SET(n) (mark[(n) >> 5] |= (1u << ((n) & 31)))

unsigned int mark[(MAX_NUM >> 5) + 1];
short dpCount[MAX_NUM + 1];

int sumOfDigits(int n) {
    int result = 0;
    while(n > 0) {
        result += n % 10;
        n /= 10;
    }
    return result;
}

void eratosthenesSieve() {
    SET(0);
    SET(1);
    for(int n = 2; n <= MAX_NUM; ++n) {
        if(n <= 999 && !GET(n)) {
            for(int i = n * n; i <= MAX_NUM; i += n)
                SET(i);
        }
        dpCount[n] = dpCount[n - 1] + (!GET(n) && !GET(sumOfDigits(n)));
    }
}

int main() {
    eratosthenesSieve();
	int N;
	scanf("%d", &N);
	while(N--) {
		int t1, t2;
		scanf("%d %d", &t1, &t2);
		printf("%d\n", dpCount[t2] - dpCount[t1 - 1]);
	}
	return 0;
}

File: 10533 Digit Primes/solution.java
import java.io.*;

public class Main {
    static int MAX_NUM = 999999;
    static boolean[] isComposite = new boolean[MAX_NUM + 1];
    static short[] dpCount = new short[MAX_NUM + 1];
	public static void main(String[] args) throws IOException {
	    isComposite[0] = isComposite[1] = true;
	    eratosthenesSieve();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
        st.nextToken();
        int N = (int)st.nval;
        StringBuilder output = new StringBuilder();
        while(N-- > 0) {
            st.nextToken();
		    int t1 = (int)st.nval;
		    st.nextToken();
		    int t2 = (int)st.nval;
			output.append(dpCount[t2] - dpCount[t1 - 1]).append('\n');
    	}
        System.out.print(output);
	}
	
	static int sumOfDigits(int n) {
        int result = 0;
        while(n > 0) {
            result += n % 10;
            n /= 10;
        }
        return result;
    }
    
    static void eratosthenesSieve() {
	    isComposite[0] = isComposite[1] = true;
        for(int n = 2; n <= MAX_NUM; ++n) {
            dpCount[n] = dpCount[n - 1];
            if(n <= 999 && !isComposite[n]) {
                for(int i = n * n; i <= MAX_NUM; i += n)
                    isComposite[i] = true;
            }
            if(!isComposite[n] && !isComposite[sumOfDigits(n)])
	            ++dpCount[n];
        }
    }
}
